# SeqOthello

__SeqOthello__ is an ultra-fast and memory-efficient indexing structure to support arbitrary sequence query against large collections of RNA-seq experiments. Taking a sequence as query input, SeqOthello returns either the total _k_-mer hits of the query sequence or the detailed presence/absence information of individual k-mers across all the indexed experiments.



## SeqOthello Installation

### System Requirements
__SeqOthello__ is tested on linux platforms with the following system settings.
 The performance is optimized for Intel CPUs with SSE4.2 support.

  * cmake >= 2.8.4
  * gcc >= 4.9.1
  * zlib >= 1.2.3

Ubuntu
```
sudo apt install cmake build-essential zlib1g-dev
```
Fedora
```
sudo yum install gcc-c++ cmake zlib-dev
```

Mac OS 10.12. Please install cmake using [brew](https://brew.sh/)
```
brew install cmake
```

### Installation

1. Clone the repository:

    ```
    git clone https://github.com/LiuBioinfo/SeqOthello.git
    ```

1. Build and install:

    ```
    cd SeqOthello/
    ./compile.sh
    ```

    After successfully build the code, you can find __SeqOthello__ toolchain at ``build/bin/``.

    ```
    ls build/bin
    ```


## Manual

### Build SeqOthello with an example.

The construction of __SeqOthello__ requires the input of a list of
k-mer files, each of which contains the set of _k_-mers extracted from
reads of the corresponding RNA-Seq experiment.
Currently the k-mer file is generated by [Jellyfish](https://github.com/gmarcais/Jellyfish) from fasta/fastq files.

For demonstration purpose, we provide an ``example/`` project  including 10 simulated pair-end RNA-seq experiments for testing. You can use the pipeline script ``example/build_and_query.sh`` to build the __SeqOthello__ map for the sample project from the ``fastq`` files and query the example transcripts included in ``example/transcripts.fa``

```
cd example/
less experiments_list.10.txt
experiment_00
experiment_01
experiment_02
experiment_03
experiment_04
experiment_05
experiment_06
experiment_07
experiment_08
experiment_09
```

1. Extract _k_-mer count using [Jellyfish](https://github.com/gmarcais/Jellyfish)

    First, use Jellyfish to generate _k_-mer files for the experiments included in ``experiments_list.10.txt`` and save them in the temporary folder ``tmp/kmers``. This setup may take about 10 seconds. For the given example, we provide a shell scripts for this step ``STEP1_Jellyfish.sh``. You may execute the script in ``example`` folder:
    ```
    ./STEP1_Jellyfish.sh
    ```

1. Convert _k_-mer files to __SeqOthello__ binary format.

The second step is to convert k-mer files to SeqOthello binary format using ``PreProcess``.

```  
   PreProcess {OPTIONS}
    Convert a Jellyfish output file to binary format supported by SeqOthello.
  OPTIONS:
      --in=[string]                     filename for the input kmer file
      --out=[string]                    filename for the output binary kmer file
      --k=[integer]                     k, the k-mer length in the input file.
      --cutoff=[integer]                cutoff. Optional value. Only k-mers with at least [cutoff]
                                        counts are kept for building SeqOthello.
      --histogram                       Use this command to generate a histogram of k-mer expression.

```
For the given example, we provide a shell scripts for this step ``STEP2_Binary.sh``. You may execute the script in ``example`` folder
    ```
    ./STEP2_Binary.sh
    ```



1. Make __SeqOthello__  group files
In the third step, binary files are grouped by the ``Group`` tool, into small subsets for further process. Generally, each group contains approximately 50 samples. Since we only have 10 samples in this example, we build two groups in tmp/grp/, each containing 5 samples.

<<<<<<< HEAD
The commandline to use ``Group`` is:

```
    Group {OPTIONS}
    Convert binary files to grouped files for SeqOthello.
  OPTIONS:
      --flist=[string]                  a file containing the filenames of the
                                        binary files. Each line of the [flist]
                                        must contain exactly one file name, e.g,
                                        xxxx.bin
      --folder=[string]                 The binary file and the corresponding
                                        xml file should be generated using
                                        PreProcess, and put in [folder] .
      --output=[string]                 output file
```

For the given example, we provide a shell scripts for this step ``STEP3_Group.sh``. You may execute the script in ``example`` folder
    ```
    ./STEP3_Group.sh
    ```
    
    
1. Build __SeqOthello__ mapping

    Now, we can build the __SeqOthello__ mapping between the entire set of _k_-mers and their experiment ids using the ``Build`` tool.
```
    Build {OPTIONS}
    Build SeqOthello!
      --flist=[string]                  a file containing the filenames of Grp
                                        files, these Grp files should be created
                                        by the Group tool. Each line should
                                        contain one file name.
      --grp-folder=[string]             a folder that contains the Grp files.
      --out-folder=[string]             a folder to put the generated SeqOthello
                                        map.
```
    You may execute the following command in ``example`` folder

    ```
    mkdir -p map

    ../build/bin/Build --flist=tmp/grp_list --grp-folder=tmp/grp/ --out-folder=map/

    ```

### Transcripts Query

__SeqOthello__ ``Query`` takes ``.fa`` files as input and can generate output in the following two format.

The file ``transcripts.fa`` contains 3 human transcript sequences and are used in the following example to demonstrate query results.

```
grep "ENST" transcripts.fa
>ENST00000431542
>ENST00000428263
>ENST00000628538
```

1. Number of _k_-mer hits per sequence

    By default, for each transcript in the query, __SeqOthello__ returns the
    total number of _k_-mer hits of each experiment in a tab-delimited table.

    ```
    ../build/bin/Query --map-folder=map/ \
    --transcript=transcripts.fa \
    --output=query.hits.txt \
    --qthread=1
    ```

    The query results has three rows, one transcript per row. The first column indicates the transcripts index matching the order in ``transcript.fa``. The rest of the columns are the query results for all the experiments in the __SeqOthello__ map.

    ```
    cat query.hits.txt
    transcript# 0   222     214     200     150     160     215     209     220     193     233
    transcript# 1   205     205     168     210     217     202     190     211     160     196
    transcript# 2   115     115     115     115     115     115     115     115     115     115
    ```


2. Detailed _k_-mer hit map per query

    If ``--detail`` is used, __SeqOthello__ may return the detailed map regarding individual _k_-mer’s presence/absence information across all the indexed experiments. This mode is limited to one transcript per query due to the large amount of output generated.

    ```
    head -1 transcripts.fa > ENST00000431542.fa

    ../build/bin/Query --map-folder=map/ \
    --transcript=ENST00000431542.fa \
    --output=ENST00000431542.hits.txt \
    --detail \
    --qthread=1
    ```

    The output has two columns. The first column is the _k_-mer sequence and the following set of columns contains the detailed hit map across all indexed experiments. We use ``+`` and ``.`` sign to indicate whether a _k_-mer is present or absence in an indexed experiment respectively.

    Here is an example output showing the first 10 _k_-mers in transcript ENST00000431542.

    ```
    head -10 ENST00000431542.hits.txt
    GTGGGAGTCGCCACCGCACCC ..........
    CGTGGGAGTCGCCACCGCACC .........+
    CCGTGGGAGTCGCCACCGCAC .........+
    GCCGTGGGAGTCGCCACCGCA .......+.+
    CGCCGTGGGAGTCGCCACCGC .......+.+
    CCGCCGTGGGAGTCGCCACCG .......+.+
    TCCGCCGTGGGAGTCGCCACC .......+.+
    ATCCGCCGTGGGAGTCGCCAC +......+.+
    CATCCGCCGTGGGAGTCGCCA +......+.+
    CCATCCGCCGTGGGAGTCGCC +....+.+.+
    ```

## SeqOthello Online

__SeqOthello__ also accommodates online features for small-batch queries. Online queries preload the entire index into memory prior to querying, and can be executed in approximately 0.09 seconds per transcript.

Use the following command to start a server on the machine, (e.g., on TCP port 3322). The service will run as a deamon.

```
../build/bin/Query \
--map-folder=map/ \
--start-server-port 3322
```

Open a new terminal, run the Client program for _k_-mer hit query.

```
../build/bin/Client \
--transcript=transcripts.fa \
--output=online.query.txt \
--kmer-hit \
--port=3322
```

## Build SeqOthello in parallel

Two of the three steps in __SeqOthello__ construction can be easily paralleled.
For example, with GNU Parallel, you can convert _k_-mer files with:

```
cat experiments_list.10.txt | \
parallel  ../build/bin/PreProcess \
--k=21 \
--cutoff=1 \
--in=tmp/kmers/{}.kmer \
--out=tmp/kmer_bins/{}.bin
```

## License
Please refer to LICENSE.TXT.


## Getting help
For questions running __SeqOthello__, please post to [SeqOthello Google Group](https://groups.google.com/forum/#!forum/seqothello)

## Known issues

- Linux systems often have a limit on the number of files that can be opened simutaneously. When there are a large number of experiment files, the Build program of SeqOthello may hit the limit. You may set it to larger numbers by using ``limit``. e.g,
``ulimit -nS 4096``

- The maximum size of k-mers supported as of now is 31.

- Each of binary file is associated with a xml file, the xml files should be put in the same folder with the binary files. Similarly, please put the associated xml files for the group files in the same folder with the group files.
