# SeqOthello

__SeqOthello__ is an ultra-fast and memory-efficient indexing structure to support arbitrary sequence query against large collections of RNA-seq experiments.



## SeqOthello Installation

### Requirements
__SeqOthello__ is tested on linux platforms with the following system settings.
 The performance is optimized for Intel CPUs with SSE4.2 support.

  * cmake >= 2.8.4
  * gcc >= 4.9.1
  * zlib >= 1.2.3
  * Jellyfish

Ubuntu
```
sudo apt install cmake build-essential zlib1g-dev
```
Fedora
```
sudo yum install gcc-c++ cmake zlib-dev
```

Mac OS 10.12. Please install cmake using [brew](https://brew.sh/)
```
brew install cmake
```

### Installation

1. Clone the repository:

    ```
    git clone https://github.com/LiuBioinfo/SeqOthello.git
    ```

1. Build SeqOthello:

    ```
    cd SeqOthello/
    ./compile.sh
    ```

    After successfully build the code, you can find __SeqOthello__ toolchain at ``build/bin/``.

    ```
    ls build/bin
    ```


## Manual

## Run SeqOthello

### Build SeqOthello with an example.

The construction of __SeqOthello__ requires the input of a list of
k-mer files, each of which contains the set of _k_-mers extracted from
reads of the corresponding RNA-Seq experiment.
Currently the k-mer file is generated by [Jellyfish](https://github.com/gmarcais/Jellyfish) from fasta/fastq files.

For demonstration purpose, we provide an ``example/`` project  including 10 simulated pair-end RNA-seq experiments for testing. You can use the pipeline script ``example/build_and_query.sh`` to build the __SeqOthello__ map for the sample project from the ``fastq`` files and query the example transcripts included in ``example/transcripts.fa``

```
cd example/
less experiments_list.10.txt
experiment_00
experiment_01
experiment_02
experiment_03
experiment_04
experiment_05
experiment_06
experiment_07
experiment_08
experiment_09
```

1. Extract _k_-mer count using [Jellyfish](https://github.com/gmarcais/Jellyfish)

    First, we use Jellyfish to generate _k_-mer files for the experiments included in ``experiments_list.10.txt`` and save them in the temporary folder ``tmp/kmers``. This setup may take a little while to run.

    Jellyfish parameters:
    * ``-m 21`` extract length 21 _k_-mer
    * ``-t 8 `` run Jellyfish with 8 threads
    * ``-L 1`` _k_-mer count cutoff

    ```
    mkdir -p tmp/kmers
    while IFS= read -r exp;
    do
        jellyfish count -s 1000M \
        -m 21 -C -t 8 -o tmp/kmers/${exp}.jf \
        fq/${exp}.R1.fastq \
        fq/${exp}.R2.fastq;

        jellyfish dump -t -L 1 \
        -c tmp/kmers/${exp}.jf \
        -o tmp/kmers/${exp}.kmer;

        rm tmp/kmers/${exp}.jf;
    done < experiments_list.10.txt
    ```

1. Convert _k_-mer files to __SeqOthello__ binary format.

    Then, we preprocess all _k_-mer files to convert them to the binary format and save them in ``tmp/kmer_bins``. ``--cutoff=N`` allows you to further filter the _k_-mers by their counts. _k_-mers with at least N counts will be kept for building __SeqOthello__. In this example, we set it to ``1`` to include all the _k_-mers. ``--k`` sets the _k_-mer length in the input file.

    ```
    mkdir -p tmp/kmer_bins

    while IFS= read -r exp
    do
        ../build/bin/PreProcess \
        --k=21 \
        --cutoff=1 \
        --in=tmp/kmers/${exp}.kmer \
        --out=tmp/kmer_bins/${exp}.bin

    done < experiments_list.10.txt
    ```

1. Make __SeqOthello__  group files

    We group the binary files into small subsets to obtain the _k_-mer occurrence maps within the each group of
    the experiments. Generally, each group contains approximately 50 samples. Since we only have 10 samples in this example, only 1 group will be generated in ``tmp/grp/``.

    Create the group list for the binary files.
    ```
    sed "s/$/\.bin/g" \
    experiments_list.10.txt \
    > tmp/binary_list.part00
    ```

    Build occurrence map for the group.
    ```
    mkdir tmp/grp

    ../build/bin/Group \
    --flist=tmp/binary_list.part00 \
    --folder=tmp/kmer_bins/ \
    --output=tmp/grp/Grp_00

    echo Grp_00 > tmp/grp_list

    ```

1. Build __SeqOthello__ mapping

    Now, we can build the __SeqOthello__ mapping between the entire set of _k_-mers and their experiment ids.

    ```
    mkdir -p map

    ../build/bin/Build \
    --flist=tmp/grp_list \
    --folder=tmp/grp/ \
    --out-folder=map/

    ```

### Transcripts Query

__SeqOthello__ ``Query`` supports _Containment_ and _Coverage_ query modes. You
can use ``transcripts.fa`` for testing. This file has 3 human transcript sequences.

```
grep "ENST" transcripts.fa
>ENST00000431542
>ENST00000428263
>ENST00000628538
```

1. __Containment Query__

    Containment Query will return the total number of _k_-mer hits for each
    experiment in a tab-delimited table.

    ```
    ../build/bin/Query --map-folder=map/ \
    --transcript=transcripts.fa \
    --output=containment.query.txt \
    --qthread=8
    ```

    The query results has 3 rows, one transcript per row. The first column indicates the transcripts index matching the order in ``transcript.fa``. The rest of the columns are the query results for all the experiments in the __SeqOthello__ map.

    ```
    cat containment.query.txt
    transcript# 0   222     214     200     150     160     215     209     220     193     233
    transcript# 1   205     205     168     210     217     202     190     211     160     196
    transcript# 2   115     115     115     115     115     115     115     115     115     115
    ```


2. __Coverage Query__

    Coverage query will return the detailed _k_-mer hits for each of _k_-mer
    in the queried transcripts. The coverage result has two columns. The first column is the _k_-mer sequence of the transcript. Column 2 use + and . signs to indicate the the k-mer hits status for the individual experiment in the SeqOthello map. A + sign indicates the k-mer exists in the experiment, whearas . indicates otherwise.

    ```
    ../build/bin/Query --map-folder=map/ \
    --transcript=transcripts.fa \
    --output=coverage.query.txt \
    --detail \
    --qthread=8
    ```

    Observe the first 10 _k_-mers in transcript ``ENST00000431542``.

    ```
    head -10 coverage.query.txt
    GTGGGAGTCGCCACCGCACCC ..........
    CGTGGGAGTCGCCACCGCACC .........+
    CCGTGGGAGTCGCCACCGCAC .........+
    GCCGTGGGAGTCGCCACCGCA .......+.+
    CGCCGTGGGAGTCGCCACCGC .......+.+
    CCGCCGTGGGAGTCGCCACCG .......+.+
    TCCGCCGTGGGAGTCGCCACC .......+.+
    ATCCGCCGTGGGAGTCGCCAC +......+.+
    CATCCGCCGTGGGAGTCGCCA +......+.+
    CCATCCGCCGTGGGAGTCGCC +....+.+.+
    ```

## SeqOthello Online

__SeqOthello__ also accommodates online features for small-batch queries. Online queries preload the entire index into memory prior to querying, and can be executed in approximately 0.09 seconds per transcript.

Use the following command to start a server on the machine, (e.g., on TCP port 3322). The service will run as a deamon.

```
../build/bin/Query \
--map-folder=map/ \
--start-server-port 3322
```

Open a new terminal, run the Client program for containment query.

```
../build/bin/Client \
--transcript=transcripts.fa \
--output=online.query.txt \
--containment \
--port=3322
```

## Build SeqOthello in parallel

Two of the three steps in __SeqOthello__ construction can be easily paralleled.
For example, with GNU Parallel, you can convert _k_-mer files with:

```
cat experiments_list.10.txt | \
parallel  ../build/bin/PreProcess \
--k=21 \
--cutoff=1 \
--in=tmp/kmers/{}.kmer \
--out=tmp/kmer_bins/{}.bin
```

## License
Please refer to LICENSE.TXT.


## Getting help
For questions running __SeqOthello__, please post to [SeqOthello Google Group](https://groups.google.com/forum/#!forum/seqothello)

## Known issues

Usually, on Linux systems, there is a limit on the number of files that can be opened simutaiously. The Build program of SeqOthello may hit such limit. To edit this limit, set ulimit to larger numbers. e.g,
``ulimit -nS 4096``
